### PR TITLE
docs: 3d mesh finetuning tutorial

### LINKED ISSUE #337: Create "Getting started" section about 3D finetuning
We have an Executor which enables Mesh encoding and should add a tutuorial for it: https://hub.jina.ai/executor/1mfkv18z

### PR DESCRIPTION
**Pull request by azayz**
closes #337

### CODE REVIEW

**Review by bwanglzu — CHANGES_REQUESTED** [review]
some general comments first, can you format code blocks, sort the imports and add quantitive evaluation results in this tutorial? i.e. pre-trained vs finetuned

File: docs/get-started/3d-mesh/index.md [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+```
+
+## Data preparation
+
+After downloading ModelNet data we need to unzip it.
+
+```shell
+pip install docarray
+pip install finetuner
+pip install trimesh
```
Comment (numb3r3) [inline docs/get-started/3d-mesh/index.md:27]: `docarray` would not be needed, since it has been included in finetuner

File: docs/get-started/3d-mesh/index.md [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+:align: center
+:width: 70%
+```
+
+```{figure} person.gif
+:align: center
+:width: 70%
+```
+
+Now let's go to loading our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.
```
Comment (bwanglzu) [inline docs/get-started/3d-mesh/index.md:136]:
```suggestion
Now let's go to load our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.
```

File: docs/get-started/3d-mesh/index.md [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+            doc.tags['finetuner_label'] = dir
+            train_docs.append(doc)
+    for file in glob.iglob(os.path.join(dir_path,'test/*.off'),recursive=True):
+            doc= Document(uri=file)
+            doc.load_uri_to_point_cloud_tensor(2048)
+            doc.tags['finetuner_label'] = dir
+            test_docs.append(doc)
+
+train_docs.save_binary('train_data_modelnet.bin')
+test_docs.save_binary('test_data_modelnet.bin')
```
Comment (bwanglzu) [inline docs/get-started/3d-mesh/index.md:168]: `dir` and `file` are python reserved keywords, besides, can we do:
```python
train = DocumentArray.from_files(..)
test = DocumentArray.from_files(..)

def preprocess(doc):
    doc.load_uri_to_point_cloud_tensor(..)
    doc.tags[..] = doc.uri
    return doc

train.apply(..)
test.apply(..)
...

```

**Review by bwanglzu — CHANGES_REQUESTED** [review]
some comments

**Comment by bwanglzu** [comment]
seems there're multiple `complete source code` in the tutorial, i think we should only have one, that means "complete"?

File: docs/get-started/3d-mesh/index.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+use it:
+
+```shell
+python executor-3d-encoder/finetune.py --model_name pointconv \
+                   --train_dataset ../train_data_modelnet.bin \
+                   --eval_dataset ../test_data_modelnet.bin \
+                   --batch_size 64 \
+                   --epochs 50 \
+                   --use-gpu
+```
```
Comment (bwanglzu) [inline docs/get-started/3d-mesh/index.md:219]: again, i don't think it's needed, brings limited value to the tutorial

File: docs/get-started/3d-mesh/index.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    hit = []
+    avg_pr = []
+    for d in da:
+        for m in d.matches[:topk]:
+            if d.uri.split('/')[-1].split('_')[0] == m.uri.split('/')[-1].split('_')[0]:
+                hit.append(1)
+            else:
+                hit.append(0)
+        avg_pr.append(average_precision(hit))
+    return np.mean(avg_pr)
```
Comment (bwanglzu) [inline docs/get-started/3d-mesh/index.md:259]: use built-in evaluator

File: docs/get-started/3d-mesh/index.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+tuned_model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
+tuned_model.load_state_dict(torch.load('checkpoints/finetuned-pointconv-d512.pth'))
+tuned_model.eval()
+
+train_da.embed(tuned_model, batch_size=128, device='cuda')
+eval_da.embed(tuned_model, batch_size=128, device='cuda')
+
+eval_da.match(train_da, limit=10)
+visualize(eval_da.matches[0].uri)
```
Comment (bwanglzu) [inline docs/get-started/3d-mesh/index.md:300]: this `visualize` function does not exist anymore I think.

**Review by bwanglzu — CHANGES_REQUESTED** [review]
i personally think its' better to make a clean/slim version of tutorial. If user want to explore indepth, they should explore the documentation

**Comment by bwanglzu** [comment]
you can rebase #338 to pass the CI

File: docs/get-started/3d-mesh/index.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+:align: center
+:width: 70%
+```
+
+```{figure} lap.gif
+:align: center
+:width: 70%
+```
+
+Now let's go to loading our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:43]: We'll go through all the files in our directory and we'll recursively read all files that end with `.off`

File: docs/get-started/3d-mesh/index.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        train.append(doc)
+
+train.save_binary('train_data_modelnet.bin')
+test.save_binary('test_data_modelnet.bin')
+
+```
+
+## Model Training
+
+Now that we have our data ready, we need a model that creates embeddings we will later use those embeddings for searching similar matches for different queries. A lot of work and research has been done in the field of 3D data embeddings there are some powerful models like [PointConv](https://arxiv.org/abs/1811.07246) which have been trained on [ShapeNet dataset](https://arxiv.org/pdf/1512.03012.pdf)
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:74]: we need a model that creates embeddings so that we can later use these embeddings

Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:74]: A lot of work and research has been done in the field of 3D data embeddings **and** there are some powerful models

File: docs/get-started/3d-mesh/index.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+In order to not reivent the wheel we will use this model. We will also use Jina's 3D Mesh Encoder which wraps these two models in an executor.
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:76]: reinvent

File: docs/get-started/3d-mesh/index.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+This executor receives Documents containing point sets data in its blob attribute, with shape (N, 3) and encodes it to embeddings of shape (D,). Now, the following pretrained models are ready to be used to create embeddings:
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:77]: and encodes them

File: docs/get-started/3d-mesh/index.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+
+- PointConv-Shapenet-d512: A PointConv model resulted in 512 dimension of embeddings, which is finetuned based on ShapeNet dataset.
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:79]: finetuned on ShapeNet dataset

File: docs/get-started/3d-mesh/index.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+- PointConv-Shapenet-d1024: A PointConv model resulted in 1024 dimension of embeddings, which is finetuned based on ShapeNet dataset.
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:80]: finetuned on ShapeNet dataset

File: docs/get-started/3d-mesh/index.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        rotation_matrix = np.array(
+            [[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]]
+        )
+        points[:, [0, 2]] = points[:, [0, 2]].dot(rotation_matrix)  # random rotation
+        points += np.random.normal(0, 0.02, size=points.shape)  # random jitter
+    return points
+```
+
+In the following code snippet we create MeshData model which encapsulates a `PointConv` model with 512 dimensions, we then load
+training and evaluation data from the binary fs we saved before. We create an optimizer and a learning rate scheduler. In this case with use
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:139]: fs = files?

File: docs/get-started/3d-mesh/index.md [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+    from torch.optim import Adam
+    from torch.optim.lr_scheduler import MultiStepLR
+    # create Adam optimizer with MultistepLR scheduler
+    optimizer = Adam(model.parameters(), lr=5e-4)
+    scheduler = MultiStepLR(optimizer, milestones=[30, 60], gamma=0.5)
+
+    return optimizer, scheduler
+```
+
+We leverage finetuner's mining strategy to imprive and finetuning the embeddings. By this we mean that the finetuner will use
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:158]: to improve and finetune

File: docs/get-started/3d-mesh/index.md [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+
+In the following code snippet we create MeshData model which encapsulates a `PointConv` model with 512 dimensions, we then load
+training and evaluation data from the binary fs we saved before. We create an optimizer and a learning rate scheduler. In this case with use
+an Adam optimizer and MultiStepLR scheduler but you can change those depending on your data and preferences.
+
+```python
+model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
+
+train = DocumentArray.load_binary('train_data_modelnet.bin') # load train dataset
+eval = DocumentArray.load_binary('test_data_modelnet.bin') # load eval dataset
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:146]: eval is reserved in Python, use val or dev

File: docs/get-started/3d-mesh/index.md [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+evaluator_pretrained = Evaluator(eval, train, pretrained_model)
+print(evaluator_pretrained.evaluate())
+```
+````
+
+The difference is shown in the tables below:
+
+| mAP@k  | pre-trained | fine-tuned |
+|--------|-------------|------------|
+| mAP@1  | 0.147       | 0.719      |
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:229]: how do you extract metrics at 1, 5, 10? The evaluator can't do that yet

File: docs/get-started/3d-mesh/index.md [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+```
+
+Finetuned PointConv:
+
+```{figure} dresser_finetuned.gif
+:align: center
+:width: 70%
+```
+
+We can clearly see that after finetuning PointConv has enhanced embeddings, and that's how you finetune a 3D model.
```
Comment (gmastrapas) [inline docs/get-started/3d-mesh/index.md:303]: `and that's how you finetune a 3D model.` insert thug life meme

File: docs/get-started/3d-mesh/index.md (hunk 16 above)
Comment (azayz) [reply docs/get-started/3d-mesh/index.md:229]: Using evaluator I get loop disconnect from AWS don't know why, that was from a custom script

File: docs/get-started/3d-mesh/index.md [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+
+import trimesh
+from docarray import Document, DocumentArray
+
+train = DocumentArray()
+test = DocumentArray()
+data_path = 'ModelNet40'
+all_docs = DocumentArray.from_files(os.path.join(data_path,'/**/**/*.off'))
+for doc in all_docs:
+    if 'test' in doc.uri :
```
Comment (makram93) [inline docs/get-started/3d-mesh/index.md:62]:
```suggestion
    if 'test' in doc.uri:
```

File: docs/get-started/3d-mesh/index.md [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+```
+
+```{figure} lap.gif
+:align: center
+:width: 70%
+```
+
+Now let's go to loading our data. We'll use `DocumentArray` from the library `docarray`. We'll go through
+all the files in our directory and we'll recursively read all files that end with .off.
+We split the files between two `DocumentArray`s depending on their uri to train and test.
```
Comment (makram93) [inline docs/get-started/3d-mesh/index.md:45]:
```suggestion
We split the files into two `DocumentArray`s depending on their uri to train and test.
```

File: docs/get-started/3d-mesh/index.md [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+    from torch.optim import Adam
+    from torch.optim.lr_scheduler import MultiStepLR
+    # create Adam optimizer with MultistepLR scheduler
+    optimizer = Adam(model.parameters(), lr=5e-4)
+    scheduler = MultiStepLR(optimizer, milestones=[30, 60], gamma=0.5)
+
+    return optimizer, scheduler
+```
+
+We leverage finetuner's mining strategy to improve and finetune the embeddings. By this we mean that the finetuner will use
```
Comment (makram93) [inline docs/get-started/3d-mesh/index.md:158]: Since we are using `TripletEasyHardMiner`, can you explain why it is called triplet and how to get the anchor-positive-negative triplets?

File: docs/get-started/3d-mesh/index.md [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+
+## Model Training
+
+Now that we have our data ready, we need a model that creates embeddings so that we can later use these embeddings for searching similar matches for different queries. A lot of work and research has been done in the field of 3D data embeddings and there are some powerful models like [PointConv](https://arxiv.org/abs/1811.07246) which have been trained on [ShapeNet dataset](https://arxiv.org/pdf/1512.03012.pdf)
+
+In order to not reinvent the wheel we will use this model. We will also use Jina's 3D Mesh Encoder which wraps these two models in an executor.
+This executor receives Documents containing point sets data in its blob attribute, with shape (N, 3) and encodes them to embeddings of shape (D,). Now, the following pretrained models are ready to be used to create embeddings:
+
+- PointConv-Shapenet-d512: A PointConv model resulted in 512 dimension of embeddings, which is finetuned on ShapeNet dataset.
+- PointConv-Shapenet-d1024: A PointConv model resulted in 1024 dimension of embeddings, which is finetuned on ShapeNet dataset.
```
Comment (makram93) [inline docs/get-started/3d-mesh/index.md:80]:
```suggestion
- PointConv-Shapenet-d512: A PointConv model, finetuned on ShapeNet dataset, resulting in 512 dimension of embeddings
- PointConv-Shapenet-d1024: A PointConv model,  finetuned on ShapeNet dataset, resulting in 1024 dimension of embeddings,.
```

**Review by makram93 — COMMENTED** [review]
Some minor comments. Overall it is good!

File: docs/get-started/3d-mesh/index.md [hunk 22]
```diff
@@ -0,0 +1,303 @@
+# Finetuning PointConv on ModelNet40 Dataset
+
+In this tutorial we will finetune PointConv on [ModelNet40 dataset](https://arxiv.org/pdf/1406.5670v3.pdf).
+The ModelNet40 dataset contains synthetic object point clouds. As the most widely used benchmark for point cloud analysis, ModelNet40 is popular because of its various categories, clean shapes, well-constructed dataset.
+
```
[the remark this answers is not in the document]
Comment (tadejsv) [reply docs/get-started/3d-mesh/index.md:5]: I think original is better than the suggestion

File: docs/get-started/3d-mesh/index.md [hunk 23]
```diff
@@ -0,0 +1,303 @@
...
+We leverage finetuner's mining strategy to improve and finetune the embeddings. By this we mean that the finetuner will use
+the mentioned label coming with every `Document` which we specified during preprocessing and will sample it as a positive example
+and then sample examples with different label as negatives.
+
+The triplet loss's goal is to pull `Documents` with same class together and those with different classes away from each other thus improving
+the final representation.
+
+```python
+tuned_model = finetuner.fit(
```
Comment (tadejsv) [inline docs/get-started/3d-mesh/index.md:166]: Can you add WandB callback for logging here?

File: docs/get-started/3d-mesh/index.md [hunk 24]
```diff
@@ -0,0 +1,303 @@
...
+tuned_model.eval()
+
+train_da.embed(tuned_model, batch_size=128, device='cuda')
+eval_da.embed(tuned_model, batch_size=128, device='cuda')
+
+eval_da.match(train_da, limit=10)
+```
+````
+
+Query:
```
Comment (tadejsv) [inline docs/get-started/3d-mesh/index.md:261]: Can you check if it is possible to create a table here, so that query, pretrained and finetuned results can be shown side to side? This would improve the presentation I think

**Review by tadejsv — CHANGES_REQUESTED** [review]

File: docs/get-started/3d-mesh/index.md (hunk 24 above)
Comment (azayz) [reply docs/get-started/3d-mesh/index.md:261]: I tried different solutions including html, I can't seem to get it done

File: docs/get-started/3d-mesh/index.md (hunk 20 above)
Comment (azayz) [reply docs/get-started/3d-mesh/index.md:158]: I don't want to give a small and insufficient answer so I added a phrase if they want to learn more I redirect them to that section in finetuner docs

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Review by gmastrapas — APPROVED** [review]
Looks good 🚀

File: docs/get-started/3d-mesh/index.md (hunk 24 above)
Comment (makram93) [reply docs/get-started/3d-mesh/index.md:275]: I don't know if you have tried this already but it may help: https://stackoverflow.com/a/24320279
I was able to show side by side in my PyCharm 
image

Comment (azayz) [reply docs/get-started/3d-mesh/index.md:275]: are they still gifs ? I tried it and it didn't work, can you perhaps share code

Comment (makram93) [reply docs/get-started/3d-mesh/index.md:275]: Yes they are still rendered as gif. Code is taken from the above link mentioned, I am attaching the snapshot.

Screenshot 2022-01-26 at 13 19 56

https://user-images.githubusercontent.com/6537525/151161605-50c54949-5007-4f31-85e6-36ee510aadb5.mov

Comment (azayz) [reply docs/get-started/3d-mesh/index.md:275]: this is what I'm getting locally
Screenshot from 2022-01-26 15-05-22
 
Screenshot from 2022-01-26 15-06-43

**Review by tadejsv — APPROVED** [review]
LGTM

**Review by makram93 — APPROVED** [review]
